### PR TITLE
Unit tests for game and author classes

### PR DESCRIPTION
**Pull request by DebyGrey**
## Catalog of my Things: Unit tests for game and author classes

In this project, I:

- wrote unit tests for game and author classes
Screenshot 2023-08-17 at 19 00 10
- All tests passed

### CODE REVIEW

File: spec/author_spec.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      expect(item).to have_received(:add_author).with(author)
+    end
+
+    it 'does not add the same item again' do
+      item = double('item')
+      allow(item).to receive(:add_author)
+      author.add_item(item)
+      expect(author.items).to eq([item])
+      expect(item).to have_received(:add_author).once
+    end
```
Comment (luigirazum) [inline spec/author_spec.rb:41]: - This test describes that the 'item' is added only one time to the 'items' collection at 'Author' class.
  - But in the test, the 'add_item' method is called just once.

> I think it would be better to make the call to 'add_item' with the same 'item' more than once so we can say the test pass.

File: spec/author_spec.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  describe '#to_s' do
+    before do
+      @author = Author.new(author_params)
+    end
+
+    it 'returns a string representation of the author' do
```
Comment (luigirazum) [inline spec/author_spec.rb:49]: ## 👏 Thanks for this test
- It is true, we have to make sure the instance is properly displayed.
> Let's add this to the other tests @iamsjunaid

File: spec/game_spec.rb
Comment (luigirazum) [inline spec/game_spec.rb:1]: Same as in the 'Author' class tests.
- If we take a look at the UML diagram, we can observe that the properties have the type of data (Integer, String, etc.).
- Don't you think that checking if a property was set with a correct type is recommended?

File: spec/game_spec.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+  describe '#initialize' do
+    it 'sets the multiplayer attribute' do
+      expect(game.multiplayer).to eq(game_params[:multiplayer])
+    end
+
+    it 'sets the last_played_at attribute' do
+      expect(game.last_played_at).to eq(Date.new(2023, 1, 1))
+    end
+  end
```
Comment (luigirazum) [inline spec/game_spec.rb:30]: - A 'Game' instance, only has two properties? ('multiplayer' and 'last_played_at')
- Isn't 'Game' a child class?

File: spec/game_spec.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    end
+
+    context 'when last played more than 2 years ago' do
+      it 'returns true' do
+        allow(Date).to receive(:today).and_return(Date.new(2025, 1, 1))
+        game.last_played_at = Date.new(2023, 1, 1)
+        expect(game.send(:can_be_archived?)).to be_truthy
+      end
+    end
+  end
```
Comment (luigirazum) [inline spec/game_spec.rb:48]: - I think it would be better to describe this test as it is in the requirements.
> - can_be_archived?() in the Game class
>   - should override the method from the parent class
>   - should return true if parent's method returns true AND if last_played_at is older than 2 years
>   - otherwise, it should return false

File: spec/spec_helper.rb
Comment (luigirazum) [inline spec/spec_helper.rb:1]: - Is this file necessary?
- You call the require_relatives here, and then in each `_spec.rb` file?

Please kindly verify this.

File: classes/game.rb [hunk 5]
```diff
@@ -8,4 +8,33 @@ def initialize(params)
     @multiplayer = params[:multiplayer]
     @last_played_at = Date.parse(params[:last_played_at])
   end
+
+  private
+
+  def can_be_archived?
+    # super && last_played_old_enough?
+    last_played_old_enough?
+  end
```
Comment (luigirazum) [inline classes/game.rb:17]: - The unit test require the file where the source code of the class is stored (its own .rb file)
- The line 15 is commented, so `can_be_archived?` method returns just the result of `last_played_old_enough?` method

Please kindly verify the requirements to fix this.

**Review by luigirazum — CHANGES_REQUESTED** [review]
## Hi @DebyGrey, I'm Luis, your code reviwer.

Good job so far! <br>👏👏👏 
There are some issues that you still need to work on to go to the next project, but you're almost there!
---
Status: ♻️ Changes Required
===


## To Highlight 🎉
- No Linters errors ✔️
- Unit tests were created. ✔️

 ## Required Changes ♻️
- [ ] Kindly take a look at the commented code.
- [ ] **_Please 👀 Check the comments below_**

## Optional suggestions ✍️
- N/A
_Every comment with the <kbd>OPTIONAL</kbd> prefix is not crucial enough to stop the approval of this PR. <br>However, I strongly recommend you take them into account as they can make your code better._

***
Cheers and Happy Coding! 👨‍💻
---
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me `@luigirazum` in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. <br>You should use the same Pull Request submitted for the first review,<br>either valid or invalid unless it is requested otherwise._** 

----
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/game_spec.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
+require 'rspec'
+require_relative '../classes/game'
+require_relative '../classes/item'
```
[the remark this answers is not in the document]
Comment (DebyGrey) [reply spec/game_spec.rb:3]: i have deleted the spec_helper file so this stays

File: spec/spec_helper.rb
Comment (DebyGrey) [reply spec/spec_helper.rb:1]: @luigirazum my test doesn't run without the file. I am not sure why but that's how I've been writing it for other projects and it works fine but if you have better methdod, kindly suggest.

File: spec/game_spec.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (DebyGrey) [reply spec/game_spec.rb:3]: @luigirazum  I have successfully removed the require_relatives

File: classes/game.rb (hunk 5 above)
Comment (DebyGrey) [reply classes/game.rb:17]: Fixed

File: .DS_Store
Comment (luigirazum) [inline .DS_Store:1]: 👀 **OPTIONAL**
Please kindly remove the `.DS_Store` it was already removed in your previous PR. What happened?

**Review by luigirazum — APPROVED** [review]
STATUS: 🟢 APPROVED!!!  🎉 🎉 🎉
===
Hi @DebyGrey, 🙋‍♂️ I'm Luis your code reviewer. 
---
### Great job !!! 👏 👏 👏 



Your project is **complete!**
Good work on making the changes requested in the previous review.

Now you can merge this branch and move on 🥇 
---
Congratulations! 🎉 🎉 🎉 :champagne: :champagne: :champagne:
===

## To Highlight 🎉
- No Linters errors ✔️
- All unit tests were implemented. ✔️

## Optional suggestions ✍️
- There is one, please kindly check the comment with 👀 
> _Every comment with the <kbd>OPTIONAL</kbd> prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

***
Happy Coding! 👨‍💻 and Good Luck 🍀 with the coming ones 🤞
---
Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me `@luigirazum` in your question so I can receive the notification, you can also reach me out on Slack using the same tag as well.

> **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._** 

----
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
